embedded-graphics/src/primitives/rounded_rectangle.rs

```diff
@@ -0,0 +1,777 @@
+
+    /// Top right corner radius
+    pub top_right: Size,
+
+    /// Bottom right corner radius
+    pub bottom_right: Size,
+
+    /// Bottom left corner radius
+    pub bottom_left: Size,
+}
+
+impl CornerRadii {
+    /// Create a new set of corner radii with all corners having equal values.
+    pub fn new_equal(radius: Size) -> Self {
+        Self {
+            top_left: radius,
+            top_right: radius,
+            bottom_right: radius,
+            bottom_left: radius,
+        }
+    }
+
+    /// Confine corner radii that are too large to a given bounding rectangle
+    fn confine(self, bounding_box: Size) -> Self {
+        // Compute free space (linear region) length along each edge. If there is no free space
+        // left, the value will be negative.
+        let free_space_top =
+            bounding_box.width as i32 - self.top_left.width as i32 - self.top_right.width as i32;
+        let free_space_right = bounding_box.height as i32
+            - self.top_right.height as i32
+            - self.bottom_right.height as i32;
+        let free_space_bottom = bounding_box.width as i32
+            - self.bottom_left.width as i32
+            - self.bottom_right.width as i32;
+        let free_space_left = bounding_box.height as i32
+            - self.top_left.height as i32
+            - self.bottom_left.height as i32;
+
+        // Find the largest overlap for all edges.  Larger overlaps result in larger negative
+        // numbers. For those curious, chained min fns is slightly faster than an array and
+        // iterator: https://godbolt.org/z/uQgUCb
+        let largest_overlap = free_space_top
+            .min(free_space_right)
+            .min(free_space_bottom)
+            .min(free_space_left);
+
+        // No overlap at all. Don't change
+        if largest_overlap >= 0 {
+            self
+        } else {
+            // Reduce each corner radius by (largest overlap / 2)
+            let reduce_by = Size::new_equal(largest_overlap.abs() as u32 / 2);
+
+            Self {
+                top_left: self.top_left.saturating_sub(reduce_by),
+                top_right: self.top_right.saturating_sub(reduce_by),
+                bottom_right: self.bottom_right.saturating_sub(reduce_by),
+                bottom_left: self.bottom_left.saturating_sub(reduce_by),
+            }
+        }
+    }
+}
+
+/// Rounded rectangle primitive.
+///
+/// Creates a rectangle with rounded corners. Corners can be circular or elliptical in shape, and
+/// each corner may have a separate radius applied to it.
+///
+/// # Overlapping corners
+///
+/// If one or more corner radii are too large to be contained along an edge of the rectangle, all
+/// corner radii will be reduced by half the largest overlap in pixels (rounded down).
+///
+/// This is similar but not identical to
+/// [how the CSS specification works](https://www.w3.org/TR/css-backgrounds-3/#corner-overlap) as it
+/// relies on floating point calculations.
```

Comment:
I'm happy to remove this if you think it's not necessary. It might be irrelevant, but also might provide some better context for the corner shrinking behaviour.

Reply:
I'll need take another look at this in my next review pass, because I'm not sure how relevant this is at the moment and I haven't really looked at the doc comments yet.